### PR TITLE
add TraceLog mapper and Zip Generator for large log and huge set of files

### PR DESCRIPTION
**Pull request by shokakucarrier**

### CODE REVIEW

File: src/main/java/org/commonjava/indy/service/archive/ContentGraphQLResource.java [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+import org.eclipse.microprofile.graphql.Name;
+import org.eclipse.microprofile.graphql.Query;
+
+import javax.inject.Inject;
+import javax.ws.rs.core.Response;
+import java.net.URL;
+
+@GraphQLApi
+public class ContentGraphQLResource {
+}
```
Comment (ligangty) [inline src/main/java/org/commonjava/indy/service/archive/ContentGraphQLResource.java]: Do we need to use graphql in this service?

File: src/main/java/org/commonjava/indy/service/archive/HealthResource.java [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+import org.eclipse.microprofile.graphql.Query;
+
+@GraphQLApi
+public class HealthResource {
+
+    @Query
+    public String hello() {
+        return "Pong";
+    }
+}
```
Comment (ligangty) [inline src/main/java/org/commonjava/indy/service/archive/HealthResource.java]: Do we need to use graphql in this service?

File: src/main/java/org/commonjava/indy/service/archive/ContentRESTResource.java [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+    @Inject
+    ContentService.ArchiveFile archiveFiles;
+
+    @GET
+    @Path("/{path: (.*)}")
+    @Produces(MediaType.APPLICATION_OCTET_STREAM)
+    public Uni<Response> getFile(@PathParam String path) {
+        File nf = new File(archiveFiles.dir, path);
+        Response.ResponseBuilder response = Response.ok((Object) nf);
```
Comment (ligangty) [inline src/main/java/org/commonjava/indy/service/archive/ContentRESTResource.java]: Do we need this class cast here? I remember the Response.ok can accept the pojo directly without classcast

File: src/main/java/org/commonjava/indy/service/archive/services/ArchiveFileService.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.inject.Inject;
+import java.io.*;
+import java.util.List;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipOutputStream;
+
+public class ArchiveFileService {
```
Comment (ligangty) [inline src/main/java/org/commonjava/indy/service/archive/services/ArchiveFileService.java:12]: Little confused here. Seems this service is not used in this pr, so what's purpose of it?

File: src/main/java/org/commonjava/indy/service/archive/services/LogService.java [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+        }
+
+        return item;
+    }
+
+
+
+    public static class TraceLog{
+        @JsonProperty("key")
+        private Map<String, String> keys;
```
Comment (ligangty) [inline src/main/java/org/commonjava/indy/service/archive/services/LogService.java]: Why is this a Map?

File: src/main/java/org/commonjava/indy/service/archive/services/LogService.java [hunk 6]
```diff
@@ -0,0 +1,276 @@
...
+                    break;
+            }
+        }
+
+        return item;
+    }
+
+
+
+    public static class TraceLog{
```
Comment (ligangty) [inline src/main/java/org/commonjava/indy/service/archive/services/LogService.java]: Recommend to extract this as a separate class but not embedded in other class. And seems this is indy's folo tracking content, so recommend to use "TrackedContent" as class name to apply same with indy. See: https://github.com/Commonjava/indy/blob/master/addons/folo/model-java/src/main/java/org/commonjava/indy/folo/model/TrackedContent.java

**Comment by yma955** [comment]
Hmm, we should groom the stories and make work divided more clearly, I see if that's part of archive generation MMENG-1344, let me decompose my done work and push separately.

File: src/main/java/org/commonjava/indy/service/archive/services/LogService.java [hunk 7]
```diff
@@ -0,0 +1,276 @@
...
+            return downloads;
+        }
+
+        public void setDownloads(List<TraceLogItem> downloads) {
+            this.downloads = downloads;
+        }
+    }
+
+
+    public static class TraceLogItem {
```
Comment (ligangty) [inline src/main/java/org/commonjava/indy/service/archive/services/LogService.java]: Recommend to extract this as a separate class but not embedded in other class. And seems this is indy's folo tracking record, so recommend to use "TrackedContentEntry" as class name to apply same with indy. See: https://github.com/Commonjava/indy/blob/master/addons/folo/model-java/src/main/java/org/commonjava/indy/folo/model/TrackedContentEntry.java

File: src/main/java/org/commonjava/indy/service/archive/services/LogService.java [hunk 8]
```diff
@@ -0,0 +1,276 @@
...
+import com.google.gson.Gson;
+import com.google.gson.stream.JsonReader;
+
+import java.io.*;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class LogService {
```
Comment (ligangty) [inline src/main/java/org/commonjava/indy/service/archive/services/LogService.java:12]: I dose not find where this service is used, so what's the purpose of it?

**Review by ligangty — CHANGES_REQUESTED** [review]

File: src/main/java/org/commonjava/indy/service/archive/ContentGraphQLResource.java (hunk 1 above)
Comment (shokakucarrier) [reply src/main/java/org/commonjava/indy/service/archive/ContentGraphQLResource.java]: I'm planning to use graphql(for small amount) and json to populate archive artifact here, and use rest to download zip file & single file

File: src/main/java/org/commonjava/indy/service/archive/services/ArchiveFileService.java (hunk 4 above)
Comment (shokakucarrier) [reply src/main/java/org/commonjava/indy/service/archive/services/ArchiveFileService.java:12]: I'm planning to place zip file related function here

File: src/main/java/org/commonjava/indy/service/archive/services/LogService.java (hunk 5 above)
Comment (shokakucarrier) [reply src/main/java/org/commonjava/indy/service/archive/services/LogService.java]: the trace log is like "key": {"id":"build-123"} and I think Map of String String can represent this model

File: src/main/java/org/commonjava/indy/service/archive/services/LogService.java (hunk 6 above)
Comment (shokakucarrier) [reply src/main/java/org/commonjava/indy/service/archive/services/LogService.java]: do I import of just copy the src

Comment (yma955) [reply src/main/java/org/commonjava/indy/service/archive/services/LogService.java]: This is done in https://github.com/Commonjava/indy-archive-service/pull/5

File: src/main/java/org/commonjava/indy/service/archive/ContentGraphQLResource.java (hunk 1 above)
Comment (yma955) [reply src/main/java/org/commonjava/indy/service/archive/ContentGraphQLResource.java]: I think local FS is enough by current design.

**Comment by yma955** [comment]
@shokakucarrier Which story you are handling for this commit, I think this will duplicate with the archive generation stuff which I'm in, I've pushed a base commit, something else will push later after finished. 
IIRC you could setup the tracking report receiving REST for your skeleton based on it. It's unnecessary to do the overlapping functions.

File: src/main/java/org/commonjava/indy/service/archive/services/LogService.java [hunk 9]
```diff
@@ -0,0 +1,276 @@
...
+
+    public TraceLog load(String TraceLogJson) {
+        Gson gson = new Gson();
+        return gson.fromJson(TraceLogJson, TraceLog.class);
+    }
+
+    public TraceLog load(InputStreamReader input){
```
Comment (shokakucarrier) [inline src/main/java/org/commonjava/indy/service/archive/services/LogService.java]: @yma88 I think maybe we can use these to fit your Track log model, and are we sure to just discard uploads?

**Comment by shokakucarrier** [comment]
> @shokakucarrier Which story you are handling for this commit, I think this will duplicate with the archive generation stuff which I'm in, I've pushed a base commit, something else will push later after finished.
> IIRC you could setup the tracking report receiving REST for your skeleton based on it. It's unnecessary to do the overlapping functions.

Okay, I'll remove some part, I was thinking about check if we are going to do some sidecar change base on the low level design of archive service, so I wrote all basic function we need.
But I'll remove some part of it to avoid collide with yours.

File: src/main/java/org/commonjava/indy/service/archive/ContentGraphQLResource.java (hunk 1 above)
Comment (shokakucarrier) [reply src/main/java/org/commonjava/indy/service/archive/ContentGraphQLResource.java]: @yma88  graphQL is a API request style, it's good for pack for uncertain amount of object request, So I'm thinking to provide a graphQL endpoint to check file status, update file, and let archive service download some file(without trace log)
are we gonna go on with it?

Comment (yma955) [reply src/main/java/org/commonjava/indy/service/archive/ContentGraphQLResource.java]: We don't need to update the files in archive, just generate, remove if necessary. We will use bcConfigId to keep the latest archive(cleanup if needed).
I suggest you could review the scoping doc/comments for the details if you are interested in the generating.

**Comment by yma955** [comment]
> Okay, I'll remove some part, I was thinking about check if we are going to do some sidecar change base on the low level design of archive service, so I wrote all basic function we need.
> But I'll remove some part of it to avoid collide with yours.

1. I think you could start with the tracking record receiving which is the part of skeleton, there are many things could do around it, how to seal or update the tracking record and report based on it. I think sidecar should implement something similar with Indy folo.
2. For archive generation which you dont need to care about(I'm in it), the based models I'll merge will resolve some of the reporting depts.
3. For the implementing on sidecar side, we need to be clear what's the inputs for sidecar, I think it's ok just the archive file, also be clear what's the archive structure constructed and how to perform build based the artifacts decompressing(I pick the content serving in sidecar, if you are interested, we could talk), plus tracking report back pass to Indy. So no much classes will be implemented based on archive service, but business coupling with it.

**Pull request merged**